Add unit aliases

jl-wynen opened the pull request:
Similar to #2808 but tries to implement a more Pythonic API. It is missing documentation but should be feature complete.

@SimonHeybrock  is this how we want to support it? Should the functions be in `sc.units`? We kept other functions out of it but I don't think that these here should be at the top level.
@g5t Does this API cover your use cases? Do you see a use for the context manager?

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +75,90 @@ def test_default_unit_for_string_is_none():
+def test_removing_undefined_alias_does_nothing(clean_unit_aliases):
+    sc.units.add_unit_alias(name='chubby', unit='100kg')
+
+    sc.units.remove_unit_alias(name='clucks')
```
Shouldn't this raise?

jl-wynen replied:
The idea was that a user would declare that they don't want a certain unit to show up. So if it wasn't even declared, removing it would just be a noop. But I see your point. And with a dict-syntax, it should definitely raise.

Also, it might be confusing. If, e.g., a user removes 'barn', it does not complain but strings will still contain it because it is built in.

SimonHeybrock replied:
> The idea was that a user would declare that they don't want a certain unit to show up.

A `pop` method could complement the `__delitem__` for this purpose.

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +75,90 @@ def test_default_unit_for_string_is_none():
+
+
+def test_unit_aliases_context_manager(clean_unit_aliases):
+    with sc.units.unit_aliases(clucks='19.3 m*A', speed='m/s'):
```
Nice to have a context manager 👍

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +75,90 @@ def test_default_unit_for_string_is_none():
+    assert 'speed' not in str(sc.Unit('m/s'))
+
+
+def test_unit_aliases_context_manager_preserves_prior_alias(clean_unit_aliases):
```
I suppose it does not if there are name clashes? The context manager really gives you a *nested* context, as opposed to a clean state?

SimonHeybrock reviewed (commented):
Why not have a dict-like interface, with `__setitem__` and `__delitem__`? This feels a bit clunky to a Python user, and I am pretty sure I will not be able to remember the function names here myself either.

jl-wynen commented:
> Why not have a dict-like interface, with `__setitem__` and `__delitem__`? This feels a bit clunky to a Python user, an I am pretty sure I will not be able to remember the function names here myself either.

and
> I suppose it does not if there are name clashes? The context manager really gives you a nested context, as opposed to a clean state?

New suggestion:
```.py
# Add
sc.units.aliases['clucks'] = sc.Unit('19.3 m*A')
sc.units.aliases['speed'] = 'm/s'

# Remove
del sc.units.aliases['speed']
del sc.units.aliases['undefined']  # raises KeyError
sc.units.aliases.clear()

# Other dict-like methods
sc.units.aliases.keys()
sc.units.aliases.values()  # maybe not so useful
sc.units.aliases.items()
# do we need more?

# context manager sets given aliases which may override previously defined ones.
# previous aliases that do not clash are preserved
# at the end, temporary aliases are removed, where there are clashes, the old ones are restored
with sc.units.aliases.temporary(dogyear='52 D'):
    ...
```

A caveat of the context manager is that it sets a global variable. So it affects all threads. (Set up and clean up are not thread safe.) The only way to avoid this is to set and clear the LLNL/Units mapping every time a unit is converted to/from string which seems overkill.

`aliases` looks a lot like a dict but it isn't one exactly. I don't know if there are any guarantees about the precedence of aliases with respect to insertion order. Of if we want to make a guarantee in case LLNL/Units does. What to you think?
This also means that the iterators don't necessarily have to iterate in insertion order.

SimonHeybrock commented:
> A caveat of the context manager is that it sets a global variable

The Pint units system solves a (probably similar) problem with a registry the you instantiate:

```python
import pint
ureg = pint.UnitRegistry()
3 * ureg.meter + 4 * ureg.cm
```

I guess our current system is too inflexible to handle this.

While it does not solve the threading issue, would it make any sense to *only* support aliases via a context? Or is this too inflexible and cumbersome for long-lived applications?

> I don't know if there are any guarantees about the precedence of aliases with respect to insertion order. Of if we want to make a guarantee in case LLNL/Units does. What to you think?
This also means that the iterators don't necessarily have to iterate in insertion order.

Not sure how you implement this, but my thought was to simply keep a dict on the Python side (which is then used to update/set/reset the underlying LLNL/units bits). Then you would get all this for free?

SimonHeybrock commented:
> > guess our current system is too inflexible to handle this.
> 
> Like I said, it is possible. But it requires setting up the underlying mapping every time `str` or `sc.Unit` are called.
> 
> > While it does not solve the threading issue, would it make any sense to only support aliases via a context? Or is this too inflexible and cumbersome for long-lived applications?
> 
> This does not play well with notebooks. I think most users will want to define some units in the beginning, possibly in an init function, and use them throughout the notebook.
> 
> > Not sure how you implement this, but my thought was to simply keep a dict on the Python side (which is then used to update/set/reset the underlying LLNL/Units bits). Then you would get all this for free?
> 
> It depends on what LLNL/Units does under the hood. We can easily guarantee that the iterators behave like for a Python dict, e.g. by using a Python dict. But that has no impact on string formatting.

Oh I get you now, you refer to the case of multiple aliases that may match when attempting string formatting? If it is only matching with a fixed scale, then this should not be an issue? Or is the system also checking for multiples of the alias, i.e., with different scale factors?

jl-wynen commented:
> Oh I get you now, you refer to the case of multiple aliases that may match when attempting string formatting? If it is only matching with a fixed scale, then this should not be an issue? Or is the system also checking for multiples of the alias, i.e., with different scale factors?

Don't know how exactly it works. But it can have multiple matches, e.g. `angstrom`, `nm`, or `nft`. The docs say that 
>  A user defined unit gets priority when converting to a string as well including when squared or cubed as part of a compound unit.

That is, `aliases['angstrom'] = '10^-10 m` will make formatting prefer `angstrom` over `nm`. But the docs do not  specify which user defined units have priority over other user defined units.

After messing around with it a little, it seems that the library only prefer aliases when the multiplier is 1.

SimonHeybrock commented:
> After messing around with it a little, it seems that the library only prefer aliases when the multiplier is 1.

👍 then I would consider ordering a non-issues. If you want to be thorough, we could add a check to ensure there are no exact duplicates in the `values()` of the alias dict?

jl-wynen commented:
Implemented the dict API.

I also implemented `repr` to show the unit in terms of its bases. But this changes how units render in notebooks. We have to at least change the code in the HTML repr to use `str` instead of `repr`. But the output of cells is also changed, which is not very nice. We can modify it by implementing one of the specialised `repr_*` functions. But this might be messy. Should we keep the old `repr` and add a separate function instead?

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +92,148 @@ def test_default_unit_for_string_is_none():
     assert units_identical(sc.Unit.from_dict(u.to_dict()), u)
+
+
+@pytest.mark.parametrize('unit_type', (str, sc.Unit))
+def test_unit_alias_overrides_str_formatting(unit_type):
+    sc.units.aliases['clucks'] = unit_type('19.3 m*A')
```
I wonder if going via strings (and units) here is the correct choice. Have you considered using a quantity, i.e., scalar variable? If we made this string based it may lead to tricky situations, e.g., with fractions "2/3 m" or something like that.

```python
sc.units.aliases['clucks'] = sc.scalar(19.3, 'm*A')
```

jl-wynen replied:
I did not consider this. And I think it is fundamentally different, because we define units here, not quantities.

I don't see why this would be tricky. We do not add any new capability for string parsing, users can already put numbers in their units. And there is likely no need to do this dynamically. Units are known ahead of time and it is possible to carefully write strings by hand in all cases.

SimonHeybrock commented on lib/python/bind_units.cpp:
```diff
@@ -87,6 +89,33 @@ units::Unit from_dict(const py::dict &dict) {
+  format_base(oss, "mol", base_units.mole());
+  format_base(oss, "cd", base_units.candela());
+  format_base(oss, "$", base_units.currency());
+  format_base(oss, "rad", base_units.radian());
```
Forgot count? Maybe this design is a bit risky?

jl-wynen replied:
Yes, I forgot it.
What is risky about this? The manual listing? I'd be happy to see a *simple* alternative in C++...

SimonHeybrock replied:
Yes, the manual listing... maybe we can at least try to avoid doing it multiple times (see above in `to_dict`)?

SimonHeybrock commented on lib/python/bind_units.cpp:
```diff
@@ -87,6 +89,33 @@ units::Unit from_dict(const py::dict &dict) {
+}
+
+std::string repr(const units::Unit &unit) {
+  assert_simple_unit(unit);
```
Hmm, raising in `repr` is not so nice. Can we return a fallback instead?

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +92,148 @@ def test_default_unit_for_string_is_none():
+
+def test_can_remove_unit_alias():
+    sc.units.aliases['clucks'] = '19.3 m*A'
+    sc.units.aliases['dogyear'] = '4492800s'
```
I didn't know you liked dogs!

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -75,3 +92,148 @@ def test_default_unit_for_string_is_none():
+
+def test_unit_aliases_context_manager_overrides_prior_alias():
+    sc.units.aliases['speed'] = 'm/s'
+    with sc.units.aliases.temporary(speed='km/s'):
```
How about "scope" as/in the name? "temporary" could mean anything.

SimonHeybrock commented on docs/reference/units.ipynb:
```diff
@@ -448,6 +471,192 @@
+   "metadata": {},
+   "source": [
+    "This result is not very useful.\n",
+    "We can make the formatter prefer angstrom by defining it as an alias of itself (or alternatively of `'10^-10 m'`):"
```
How will it format 'meter' in this case?

jl-wynen replied:
Still as `m`:
```.py
import scipp as sc
sc.units.aliases['angstrom'] = 'angstrom'
for p in range(12):
    print(sc.Unit(f'{10**-p}*m'))
```
```
m
dm
cm
mm
0.0001m
1e-05m
µm
9.99999999999999955e-08m
1e-08m
nm
angstrom
9.99999999999999939e-12m
```
This case is unaffected because LLNL/Units already knows about angstrom. The alias is more useful here:
```.py
import scipp as sc
sc.units.aliases['angstrom'] = 'angstrom'
for p in range(12):
    print(sc.Unit(f'{10**(2*p)}*us/m^2'))
```
left without alias, right with alias:
```
s/km^2          s/km^2
s/hectare       s/hectare
s/are           s/are
s/m^2           s/m^2
100T/V          as/angstrom^2
s/cm^2          s/cm^2
MT/V            MT/V
100000000T/V    ps/angstrom^2
10000000000T/V  10000000000T/V
TT/V            TT/V
1e+20s/km^2     µs/angstrom^2
1e+19ms/m^2     1e+19ms/m^2
```

SimonHeybrock commented on docs/reference/units.ipynb:
```diff
@@ -448,6 +471,192 @@
+    "## Prioritizing Units in String Formatting\n",
+    "\n",
+    "When dealing with crystals or molecules, it is often convenient to use angstrom as a unit.\n",
+    "But by default, string formatting tends to prefer different bases in composite units.\n",
```
MAybe we should make clear that 'angstrom' on its own works well and nothing has to be done... only int the combination with other units it may cause trouble?

SimonHeybrock commented on docs/reference/units.ipynb:
```diff
@@ -448,6 +471,192 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Alternatively, we can pass `'2.4 m'` instead of a variable.\n",
+    "\n",
+    "The string formatter now prioritizes 'horse' over 'm' when appropriate:"
```
It might be confusing for users that are not aware of units scales... can you
1. give an example (or all examples?) of formatting *variables* with this unit, so they see the value in the variable as well as the scale factor.
2. Show as counterexample that `sc.scalar(2.4, 'm')` does *not* format as `'horse'`?

SimonHeybrock reviewed (commented):
Notes on the notebook:

SimonHeybrock commented on docs/reference/units.ipynb:
```diff
@@ -630,7 +634,160 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sc.units.aliases['dogyear'] = sc.scalar(4492800, unit='s')"
```
I don't want to nitpick too much, but wouldn't it be more useful to give a real example? Why not go with the lattice spacing application?

jl-wynen replied:
Because everyone has some idea of what dog years are. An application specific units needs more explanation.

SimonHeybrock commented on src/scipp/units/__init__.py:
```diff
@@ -31,9 +30,171 @@
+    def __setitem__(self, alias: str, unit: Union[str, Unit, Variable]):
+        """Define a new unit alias."""
+        multiplier, unit = _unit_and_multiplier(unit)
+        if unit in self.values():
+            raise ValueError(f"There already is an alias for unit '{unit!r}'")
```
Is this check correct? Shouldn't it check *with* the multiplier? This looks like it would prevent defining an alias for different lattice spacings, e.g., in different directions.

jl-wynen replied:
Right. Missed this when adding the separate multiplier.

SimonHeybrock commented on lib/units/unit.cpp:
```diff
@@ -208,11 +208,8 @@ bool identical(const Unit &a, const Unit &b) {
-  llnl::units::addUserDefinedUnit(name, u);
-  return Unit(u);
+void add_unit_alias(const std::string &name, const Unit &unit) {
+  llnl::units::addUserDefinedUnit(name, unit.underlying());
```
Hmm, do you have a test with a quantity *and* an existing multiplier? Do the two factors get combined correctly?

jl-wynen replied:
See your other comment. The 3rd case does that

SimonHeybrock commented on tests/unit_test.py:
```diff
@@ -150,6 +155,10 @@ def test_defining_conflicting_alias_raises():
     sc.units.aliases['speed'] = 'm/s'
     with pytest.raises(ValueError):
         sc.units.aliases['fastness'] = 'm/s'
+    with pytest.raises(ValueError):
+        sc.units.aliases['fastness'] = '100*cm/s'
+    with pytest.raises(ValueError):
+        sc.units.aliases['fastness'] = sc.scalar(1000, unit='mm/s')
```
Not sure I get this test:
- Overriding alias (same name) should be fine?
- Different alias for same unit should raise.
- I can't see a test for different units with same base dims (but different scales) with different aliases. Having alias es for `1.2 m` and `1.4 m` at the same time should be possible, no?

jl-wynen replied:
> Not sure I get this test:
> 
>     * Overriding alias (same name) should be fine?

Yes, see `test_can_override_alias`

> 
>     * Different alias for same unit should raise.

This is what this test checks.

> 
>     * I can't see a test for different units with same base dims (but different scales) with different aliases. Having alias es for `1.2 m` and `1.4 m` at the same time should be possible, no?

True, I can add that.

SimonHeybrock commented on docs/reference/units.ipynb:
```diff
@@ -448,6 +471,350 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Unit Aliases\n",
```
```suggestion
    "## Unit Aliases\n",
```
